### PR TITLE
CDP partial implementation

### PR DESCRIPTION
**Pull request by yotamdaniel**
Whats in this PR:

 - Implementation of Withdraw and deposit methods in CDP
 - Implementation of create and destroy position methods in CDP
 - Implementation of global collateral methods in CDP
 - Implementation CDP and various getters
 - tests for all of the above 

- add .mocharc.json for easier VSCode integration - feel free to remove if needed or bothersome

Whats not in PR:
  - paying interest method implementation 
  - rate calculator methods
  - No address based implementation of CDP. Only dealing with NFTs ID at this point

### CODE REVIEW

**Review by tomerXarchimedes — CHANGES_REQUESTED** [review]
Some of the comments might be on older commits and you might already fixed them
I think the big missing things are
* Handling borrow from an address (not just NFT) 
* relevant tests (probably split most tests to NFT and address)

File: contracts/CDPosition.sol [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+contract CDPosition {
+    struct cdp {
+        uint256 oUSDPrinciple; // Amount of OUSD originally deposited by user
+        uint256 oUSDInterestEarned; // Total interest earned (and rebased) so far
+        uint256 oUSDTotal; // Principle + OUSD acquired from selling borrowed lvUSD + Interest earned
+        uint256 lvUSDBorrowed; // Total lvUSD borrowed under this position
+        bool firstCycle; // to prevent quick "in and out", we don't credit interest to a position at first the interest payment cycle
+    }
+
+    mapping(uint256 => cdp) private nftCDP;
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply contracts/CDPosition.sol:17]: We need another address<>CDP mapping. Its not in this pull request though as I want to first focus on NFTs. Adding address should be a straightforward refactor. 

Will change private to internal

File: contracts/CDPosition.sol [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    /// @dev delete entry in CDP --if-- lvUSD borrowed balance is zero
+    ///
+    /// @param nftID entry address to delete
+    function deletePosition(uint256 nftID)
+        external
+        nftIDMustExist(nftID)
+        canDeletePosition(nftID)
+    {
+        /// Set all values to default. Not way to remove key from mapping in solidity
+        nftCDP[nftID] = cdp(0, 0, 0, 0, false);
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply contracts/CDPosition.sol:40]: Right, changing this to `delete nftCDP[nftID]`

File: contracts/CDPosition.sol [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    /// @dev update deposited OUSD in position. This method removed a delta to existing deposited value
+    /// @param nftID NFT position to update
+    /// @param oUSDAmountToWithdraw amount to remove to position's existing deposited sum
+    function withdrawOusdFromPosition(
+        uint256 nftID,
+        uint256 oUSDAmountToWithdraw
+    ) external nftIDMustExist(nftID) {
+        nftCDP[nftID].oUSDTotal += oUSDAmountToWithdraw;
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply contracts/CDPosition.sol]: I'm now realizing this commit has a slightly older version with some bugs I already weeded out. Newest commit should fix those including this one

File: test/CDPosition.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+    const NFT_ID_1 = 1234;
+    const NFT_ID_2 = 67889;
+    const BASIC_OUSD_PRINCIPLE = getEighteenDecimal(1000000)
+
+    beforeEach(async function () {
+        let contract = await ethers.getContractFactory("CDPosition");
+        [owner, addr1, addr2, ...addrs] = await ethers.getSigners();
+        cdp
+            = await contract.deploy();
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply test/CDPosition.js:37]: No, its an unruly auto formatter - leaving it for now but expect a better styling guide soon with an auto formatter config everyone on the team can share

File: test/CDPosition.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        expect(await cdp.getOUSDTotal(nftID)).to.equal(total);
+        expect(await cdp.getLvUSDBorrowed(nftID)).to.equal(borrowed);
+        expect(await cdp.getFirstCycle(nftID)).to.equal(firstCycle);
+    }
+
+    let cdp;
+
+    const NFT_ID_1 = 1234;
+    const NFT_ID_2 = 67889;
+    const BASIC_OUSD_PRINCIPLE = getEighteenDecimal(1000000)
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply test/CDPosition.js]: Added

File: test/CDPosition.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        it("Should create position", async function () {
+            // create a new position
+            await cdp.createPosition(NFT_ID_1
+                , BASIC_OUSD_PRINCIPLE)
+
+            // validate CDP Values
+            await validateCDP(NFT_ID_1
+                , BASIC_OUSD_PRINCIPLE, 0, BASIC_OUSD_PRINCIPLE, 0, true)
+        });
+        it("Should not create position on existing address", async function () {
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply test/CDPosition.js:50]: Changing to NFT

File: hardhat.config.js [hunk 7]
```diff
@@ -19,12 +25,12 @@ task("accounts", "Prints the list of accounts", async (taskArgs, hre) => {
...
-//  }
+  //  networks: {
+  //    hardhat: {
+  //      forking: {
+  //        url: "https://eth-mainnet.alchemyapi.io/v2/<key>",
+  //        blockNumber: 11095000
+  //      }
+  //    }
+  //  }
 };
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply hardhat.config.js:30]: Reverting

File: test/CDPosition.js [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            await cdp.depositOUSDtoPosition(NFT_ID, depositOUSDAmount)
+            await cdp.depositOUSDtoPosition(NFT_ID_SECONDARY, depositOUSDAmount)
+        })
+        
+        it("Should pay interest to positions", async function () {
+            await cdp.payInterestToPositions(interestToPay)
+            let nftIdMainOUSDInterestEarned = await cdp.getOUSDInterestEarned(NFT_ID);
+            let nftIdSecondaryOUSDInterestEarned = await cdp.getOUSDInterestEarned(NFT_ID_SECONDARY);
+            console.log("nftIdMainOUSDInterestEarned %s", nftIdMainOUSDInterestEarned)
+            console.log("nftIdSecondaryOUSDInterestEarned %s", nftIdSecondaryOUSDInterestEarned)
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply test/CDPosition.js]: removing

File: contracts/CDPosition.sol [hunk 9]
```diff
@@ -11,14 +11,31 @@ contract CDPosition {
...
     }
-    
+
     uint256 private globalCollateralRate;
 
     mapping(uint256 => cdp) private nftCDP;
+    /// late add on to be able to iterate 
+    /// TODO : remove position when deleting, add position when creating. Maybe we want to check
+    uint256 public totalEntries = 0;
+    uint256[] createdNFtPositions;
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply contracts/CDPosition.sol]: Yes, double checking all other access qualifiers

File: hardhat.config.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply hardhat.config.js]: Removing

File: contracts/CDPosition.sol [hunk 10]
```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.4;
+
+import "hardhat/console.sol";
+
+contract CDPosition {
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply contracts/CDPosition.sol:11]: Adding now

File: contracts/CDPosition.sol [hunk 11]
```diff
@@ -82,6 +84,25 @@ contract CDPosition {
...
+    // /// @param nftID 
+    // function amountOfLvUSDAvailableToBorrow(uint256 nftID) view public; -- need to know collateral rate  
+
+    /// @dev update collateral rate 
+    ///
+    /// @notice Max lvUSD that can be minted for 1 OUSD
+    ///
+    /// @param ratio new ratio to set as collateral
+    function changeCollateralRate(uint256 ratio) external {
+        globalCollateralRate = ratio;
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply contracts/CDPosition.sol:102]: Changing all to rate

File: test/CDPosition.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            await cdp.repayLvUSDToPosition(NFT_ID_1, getEighteenDecimal(1000))
+            await validateCDP(NFT_ID_1, BASIC_OUSD_PRINCIPLE, 0, BASIC_OUSD_PRINCIPLE, getEighteenDecimal(9000), true)
+        })
+
+        it("Should not mark down repayed lvUSD if not enough borrowed lvUSD", async function () {
+            await expect(cdp.repayLvUSDToPosition(NFT_ID_1, getEighteenDecimal(1000)))
+                .to.be.revertedWith("lvUSD Borrowed amount must be greater than amount to repay")
+        })
+
+    })
```
[the remark this answers is not in the document]
Comment (yotamdaniel) [reply test/CDPosition.js:149]: There can be no end to testing, this is the base layer of tests needed for developing the CDP using TDD.

I suggest we add more tests in test bashes and ongoing

**Review by tomerXarchimedes — APPROVED** [review]

**Pull request merged**